### PR TITLE
124 stop rule reporting

### LINKED ISSUE #94: Add additional statistics to the analysis of simulation output
For the final protocol we analyze and present statistics that are currently not covered by summary method in crmPack. In addition, the summary method in crmPack is tailored for the NextBestNCRM-class but fit not very well for other next best dose approaches, e.g. smallest distance to target we want to add to crmPack.

Here is an example what we have presented in past in the protocol:

image

In the future we will change this reporting, e.g. we will add the selection probability and probably exchange min/max by the 10% and 90% quantiles as currently presented in crmPack. However, for assessment of the operation characteristics in general it will still be useful to look also at min/max. 

Some of the statistics that are presented cannot be derived from the current S4 object of class simulation in crmPack (e.g. median MTD) or are not easy to derive (e.g. reason for stopping).
 
Therefore, changes to existing functions in crmPack may be necessary. In the following we want to give more details using the stopping rules as an example.

In principle, the current S4 object of class simulation in crmPack contains the stopReasons as list. The individual results why a study has stopped, is captured as text and in one can identify which rule is triggered. 
As the stopping text contains detailed information that vary between simulated studies and in case the rule is triggered or not, it is not straight forward to be used for further reporting.

We tried the following ways to make the name of the stopping rule and the status of a stopping rule available to the simulation output object:
1)	Via the text that is returned by the stopping rule:
A slot ‘label’ is added to the stopping rule. The label (stopping@label) and  the value of the variable doStop are added to the returned text, so that it looks like
… paste(doStop, ' : ', stopping@label, ' : ', …text as before….)
2)	Via the simulation function:
The simulation function is updated so that the name of the stopping rule and the status of the stopping rule is written to the simulation output.

The advantage of implementation 1 is that a label is introduced that can be directly used for reporting. The disadvantage is that the output of any stopping rule must be changed to follow the new output text pattern so that it can be analyzed automatically.

The advantage of implementation 2 is that the individual stopping rules do not need to be adapted. A possible disadvantage is that the simulation function needs to be adapted.

Given the goal that additional statistics can be derived from the simulation output (e.g. median MTD) a change of the simulation function seems to be necessary in any case.

An Idea of implantation without affecting the current reporting mechanism, could be to create an additional (new) design class, that is picked up by a revised simulation function and reporting function. This may be needed anyway, when another mechanism for selecting the next best dose is used (no interval around the target toxicity is used).

We like to get thought if this makes sense at all and if so, how to implement this in the best way into the package. Probably there better ways to archive the reporting goal.

### PR DESCRIPTION
**Pull request by ClaraBeckBayer**
# Pull Request

Pull request for new branch for stopping rule reporting.

Fixes #94 (partially)

### CODE REVIEW

File: R/Rules-class.R [hunk 1]
```diff
@@ -1519,7 +1544,11 @@ StoppingCohortsNearDose <- function(nCohorts = 2L,
 #' @note Typically, end users will not use the `.DefaultStoppingCohortsNearDose()` function.
 #' @export
 .DefaultStoppingCohortsNearDose <- function() { # nolint
-  StoppingCohortsNearDose(nCohorts = 3L, percentage = 0.2)
+  StoppingCohortsNearDose(
+    nCohorts = 3L,
+    percentage = 0.2,
+    report_label = NA_character_
```
Comment (danielinteractive) [inline R/Rules-class.R:1550]:
```suggestion
    percentage = 0.2
```
not needed explicitly here, it is the default for that argument anyway. similar below for other default constructors

File: R/Rules-class.R [hunk 2]
```diff
@@ -2221,14 +2534,16 @@ StoppingList <- function(stop_list, summary) {
...
 #'
 #' @param stop_list (`list`)\cr see slot definition.
-#'
+#' @param report_label (`string`) \cr see slot definition.
 #' @export
 #' @example examples/Rules-class-StoppingAll.R
 #'
-StoppingAll <- function(stop_list) {
-  .StoppingAll(stop_list = stop_list)
+StoppingAll <- function(stop_list, report_label = character(0)) {
```
Comment (danielinteractive) [inline R/Rules-class.R:2541]: why do we have here `character(0)` as default instead of `NA_character_` above?

File: R/Design-methods.R [hunk 3]
```diff
@@ -1063,6 +1089,9 @@ setMethod("simulate",
       ## the reasons for stopping
       stopReasons <- lapply(resultList, "[[", "stop")
 
+      ## for dual simulations as it would fail in summary otherwise (for dual simulations reporting is not implemented)
+      stop_report <- matrix(TRUE, nrow = nsim)
```
[the remark this answers is not in the document]
Comment (ClaraBeckBayer) [reply R/Design-methods.R:1093]: No, because this matrix is not supposed to contain missing values: 
image

File: R/Design-methods.R [hunk 4]
```diff
@@ -398,13 +398,33 @@ setMethod("simulate",
...
+              value <- stopit_tree[1]
+              names(value) <- label
+              stopit_unpacked <- append(stopit_unpacked, value)
+              if (is.null(attr(stopit_tree, "individual"))) {
+                stopit_unpacked
+              } else {
+                append(stopit_unpacked, lapply(attr(stopit_tree, "individual"), unpack_stopit))
+              }
+            }
+          }
```
[the remark this answers is not in the document]
Comment (ClaraBeckBayer) [reply R/Design-methods.R:425]: Okay, I added it to the helpers.R file, is this the right place?

File: R/Rules-class.R (hunk 2 above)
Comment (ClaraBeckBayer) [reply R/Rules-class.R:2541]: Changed this to NA for StoopingAll and StoppingAny too.

File: R/Design-methods.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (danielinteractive) [reply R/Design-methods.R:425]: yep that works thanks!

File: R/helpers.R [hunk 5]
```diff
@@ -1044,3 +1044,30 @@ h_find_interval <- function(..., replacement = -Inf) {
...
+
+#'unpack stopping rules and return list
+#'
+#'@description
+#'
+#'recursively unpack nested stopping rules logical value and label given
+#'
+#'@param stopit_tree object from simulate method
+
+h_unpack_stopit <- function(stopit_tree) {
```
Comment (danielinteractive) [inline R/helpers.R:1058]: can you please add unit tests covering all if/else paths for this function?

Comment (ClaraBeckBayer) [reply R/helpers.R:1056]: Robert optimized the function so there is only one if condition left now.

File: R/Simulations-methods.R [hunk 6]
```diff
@@ -977,6 +977,17 @@ setMethod("show",
         dfNames = dfNames
       )
 
+      # report stopping rules
+      # Report individual stopping rules with non-<NA> labels.
+      stop_pct <- colMeans(object@stop_report) * 100
```
Comment (danielinteractive) [inline R/Simulations-methods.R:982]: Can you please add tests that check whether this reporting works as expected? including potential edge cases. Since this is the main feature of this PR

**Review by danielinteractive — APPROVED** [review]
Just one minor comment then good to merge

File: R/Simulations-methods.R [hunk 7]
```diff
@@ -979,8 +978,11 @@ setMethod("show",
 
       # report stopping rules
       # Report individual stopping rules with non-<NA> labels.
-      stop_pct <- colMeans(object@stop_report) * 100
-      stop_pct_to_print <- stop_pct[!is.na(names(stop_pct))]
+      # stop_pct <- colMeans(object@stop_report) * 100
+      # stop_pct_to_print <- stop_pct[!is.na(names(stop_pct))]
```
[the remark this answers is not in the document]
Comment (robertadamsbayer) [reply R/Simulations-methods.R]: Solved
